ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -672,34 +672,52 @@ func (r *RayClusterReconciler) updateStatus(instance *rayiov1alpha1.RayCluster)
 		}
 	}
 
+	if err := r.updateEndpoints(instance); err != nil {
+		return err
+	}
+
+	timeNow := metav1.Now()
+	instance.Status.LastUpdateTime = &timeNow
+	if err := r.Status().Update(context.Background(), instance); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *RayClusterReconciler) updateEndpoints(instance *rayiov1alpha1.RayCluster) error {
+	// TODO: (@scarlet25151) For now there are be several K8s Services for
+	// one RayCluster. We filter for head Services with label selectors and pick the first one.
+	// We may need use the Get method to select by name in the future.
 	rayHeadSvc := corev1.ServiceList{}
-	filterLabels = client.MatchingLabels{
+	filterLabels := client.MatchingLabels{
 		common.RayClusterLabelKey:  instance.Name,
 		common.RayNodeTypeLabelKey: "head",
 	}
-	// TODO: (@scarlet25151) for now there would be several kubernetes serivces related to
-	// one raycluster, we have used the label to select the headservice and pick the first one.
-	// we may need use Get method to select by name.
 	if err := r.List(context.TODO(), &rayHeadSvc, client.InNamespace(instance.Namespace), filterLabels); err != nil {
 		return err
 	}
+
 	if len(rayHeadSvc.Items) != 0 {
 		svc := rayHeadSvc.Items[0]
 		if instance.Status.Endpoints == nil {
 			instance.Status.Endpoints = map[string]string{}
 		}
 		for _, port := range svc.Spec.Ports {
 			if len(port.Name) == 0 {
-				r.Log.Info("updateStatus", "service port name is empty", port)
+				r.Log.Info("updateStatus", "service port's name is empty. Not adding it to RayCluster status.endpoints", port)
 				continue
 			}
-			instance.Status.Endpoints[port.Name] = fmt.Sprintf("%d", port.NodePort)
+			if port.TargetPort.IntVal != 0 {
+				instance.Status.Endpoints[port.Name] = fmt.Sprintf("%d", port.TargetPort.IntVal)
+			} else if port.TargetPort.StrVal != "" {
+				instance.Status.Endpoints[port.Name] = port.TargetPort.StrVal
+			} else {
+				r.Log.Info("updateStatus", "service port's targetPort is empty. Not adding it to RayCluster status.endpoints", port)
+			}
 		}
-	}
-	timeNow := metav1.Now()
-	instance.Status.LastUpdateTime = &timeNow
-	if err := r.Status().Update(context.Background(), instance); err != nil {
```

Comment:
I think this update is still needed because there are other status need to be update at this time.

Reply:
I update the status [here](https://github.com/ray-project/kuberay/pull/383/files#diff-72ecc3ca405f1e828187748d4f1ec8160bccffa2a4f84a364cd7a94a78e1adb9R681). Is that OK? Or do I need to update the status multiple times like once before I try to update endpoints?

Reply:
I see,I think you are right to update the instance at once when all the field are populated.